src/GraphQL.MemoryCache/MemoryDocumentCache.cs

```diff
@@ -62,17 +63,30 @@ protected virtual MemoryCacheEntryOptions GetMemoryCacheEntryOptions(string quer
         }
 
         /// <inheritdoc/>
-        public virtual Document? this[string query]
+        public virtual void Dispose()
         {
-            get => _memoryCache.TryGetValue<Document>(query, out var value) ? value : null;
-            set => _memoryCache.Set(query ?? throw new ArgumentNullException(nameof(query)), value, GetMemoryCacheEntryOptions(query));
+            if (_memoryCacheIsOwned)
+                _memoryCache.Dispose();
         }
 
         /// <inheritdoc/>
-        public virtual void Dispose()
+        public virtual ValueTask<Document?> GetAsync(string query) =>
+            new(_memoryCache.TryGetValue<Document>(query, out var value) ? value : null);
+
+        /// <inheritdoc/>
+        public virtual ValueTask SetAsync(string query, Document value)
```

Comment:
```csharp
        public virtual ValueTask SetAsync(string query, Document value)
        {
            if (value == null)
                _memoryCache.Remove(query);
            else
                _memoryCache.Set(query, value, GetMemoryCacheEntryOptions(query));

            return default;
        }
```

and remove `public virtual ValueTask RemoveAsync(string query)`

Reply:
Then we should revert back to `Document?`

Reply:
Either the argument is nullable or it isn’t. Having it defined as nullable does not hurt any other code.

Reply:
> Either the argument is nullable or it isn’t. Having it defined as nullable does not hurt any other code.

Actually, the setter with a null value is expected to set a real cache item as `{key} => null` (not to remove an existing item). If the RemoveAsync method is considered no-need, we just do not provide any cache removing way and keep the setter as a single responsibility method.

Reply:
```csharp
        public virtual Document? this[string query]
        {
            get => _memoryCache.TryGetValue<Document>(query, out var value) ? value : null;
            set => _memoryCache.Set(query ?? throw new ArgumentNullException(nameof(query)), value, GetMemoryCacheEntryOptions(query));
        }
```

> If the argument is passed as non-nullable, it should be expected that it may throw an ArgumentNullException if it is null. We should not disguise the method intent behind incorrect NRT annotations.

@Shane32 By the way, the previous design allows setting a null cache value.

If we do this change, projects that need the caches removing feature may be hurt.

Reply:
> the setter with a null value is expected to set a real cache item as {key} => null (not to remove an existing item).

Why "expected". It's just an implementation detail. The client of this method gets identical behavior regardless of whether the cache keeps `null` values or not.

Reply:
Moreover, the storage of null values implies keeping **keys**, which affects the overall memory consumption.

Reply:
> the previous design allows setting a null cache value.

Actually it doesn't.  See the null-coalescing operator where it throws an exception if a null value is passed in.

```cs
_memoryCache.Set(query ?? throw new ArgumentNullException(nameof(query))
```

> Why "expected"? It's just an implementation detail. The client of this method gets identical behavior regardless of whether the cache keeps null values or not.

I agree.

Reply:
Perhaps for the simplest implementation, let's not support removing key values at all.  If a user requests a change, we can perform the following as a non-breaking change:

1. Change the NRT annotation on `IDocumentCache.SetAsync`'s `document` argument to indicate it is nullable.
2. Rewrite the `MemoryDocumentCache.SetAsync` code to clear an entry if `null` is passed in.

(Note: adding another member, such as `RemoveAsync`, to an interface is a breaking change.)

Reply:
This suggestion will also match up with the existing implementation, which it throws an `ArgumentNullException` if trying to set a value to the cache which is `null`.

Reply:
Right, I see.

Reply:
> Perhaps for the simplest implementation, let's not support removing key values at all. If a user requests a change, we can perform the following as a non-breaking change:
> 
> 1. Change the NRT annotation on `IDocumentCache.SetAsync`'s `document` argument to indicate it is nullable.
> 2. Rewrite the `MemoryDocumentCache.SetAsync` code to clear an entry if `null` is passed in.
> 
> (Note: adding another member, such as `RemoveAsync`, to an interface is a breaking change.)

Ok then let's go with this @gdlcf88 if we are all agreed.

Reply:
@Shane32 @sungam3r

Done.

https://github.com/graphql-dotnet/graphql-dotnet/pull/2758/commits/384e3c0d2ad67a70e8fb08395e422f6490bde966

Reply:
Well, what we wrote is to ** not ** support removing keys at all -- and change it later if someone requests it.  (lol)

Reply:
https://github.com/graphql-dotnet/graphql-dotnet/pull/2758/commits/ce8b43aa1f5da5a03ed53726a965891e33178ada